File: Daemon/Utils/HttpServer.cs
using Daemon.Utils;
using System;
using System.Collections.Generic;
using System.Net;
using System.Threading;

namespace Daemon.Utils
{
    class HttpServer
    {
        public HttpServer(string uri)
        {
            listener = new HttpListener();

            listener.Prefixes.Add(uri);
        }

        public void Serve()
        {
            listener.Start();

            while (true)
            {
                ThreadPool.QueueUserWorkItem(ProcessRequest, listener.GetContext());
            }
        }

        public void InstallHandler(string path, Action<HttpListenerContext> handler)
        {
            handlers[path] = handler;
        }

        private void ProcessRequest(object o)
        {
            HttpListenerContext context = (HttpListenerContext)o;
            try
            {
                var handler = handlers[context.Request.Url.AbsolutePath];
                handler.Invoke(context);
                context.Response.Close();
                context = null;
            }
            catch (KeyNotFoundException)
            {
                if (context != null)
                    context.Response.SendError(HttpStatusCode.NotFound);
            }
            catch (Exception)
            {
                if (context != null)
                    context.Response.SendError(HttpStatusCode.InternalServerError);
            }
        }

        private HttpListener listener;
        private Dictionary<string, Action<HttpListenerContext>> handlers = new Dictionary<string, Action<HttpListenerContext>>();
    }
}

File: Daemon/HttpServer.fs
namespace Daemon

open System.Net
open System

type HttpServer(uri:string) =

    let listener = new HttpListener()
    let pathPrefix = (new Uri(uri.Replace("+", "localhost"))).AbsolutePath
    do  
        listener.Prefixes.Add(uri)
    let mutable stopRequested = false

    let ExtractCommandPath (uriPath:string) =
        if not (uriPath.StartsWith pathPrefix) then ""
        else uriPath.Substring(pathPrefix.Length)

    let handleRequest (context:HttpListenerContext) handlers = async {
        try
            let handler = handlers (ExtractCommandPath context.Request.Url.AbsolutePath)
            handler context
            context.Response.Close()
        with
            | _ -> Http.SendError context.Response HttpStatusCode.InternalServerError
     }

    member this.Serve handlers = 
        listener.Start()

        stopRequested <- false

        let asynctask = Async.FromBeginEnd(listener.BeginGetContext,listener.EndGetContext)
        async {
            while not stopRequested do
                try
                    let! context = asynctask
                    Async.Start (handleRequest context handlers)
                with
                    | :? HttpListenerException -> ()
        } |> Async.RunSynchronously 

    member this.Stop = 
        stopRequested <- true
        listener.Stop()

    member this.SendNotFound (context:HttpListenerContext) =
        Http.SendError context.Response HttpStatusCode.NotFound
    
